### ISSUE TITLE
CORS headers not sent when an exception is thrown

### ISSUE THREAD

**Issue by sebj54** [mevdschee/php-crud-api#645]
Hi there!

I found a very strange bug today. When I throw an exception in the `beforeHandler` hook, the CORS headers are not set in the response. In my opinion, this is a different problem than this one: https://github.com/mevdschee/php-crud-api/issues/523.

Here are my tests: 
 - hit `/records/test` (table test exists): 200 OK, with CORS headers
 - hit `/records/does_not_exist` (table does not exist): 404 NOT FOUND, with CORS headers
 - hit `/records/test` (with `beforeHandler` like in the following example): 500 SERVER ERROR, **without** CORS headers

`beforeHandler` used for the last test:
```php
'customization.beforeHandler' => function() {
    throw new \Exception('Test error');
},
```

This bug is annoying because I can't examine the error message in my apps (in JS) to display a custom error message.

I think this also can happen with other hooks such as authorization or sanitation handlers.

I would like to make a PR for this and started to work but I have to make architectural decisions. The problem is in the `Api` class, in the `handle` function (see code comments):

```php
public function handle(ServerRequestInterface $request): ResponseInterface
{
    $response = null;
    try {
        // Here, all middlewares are processed
        $response = $this->router->route($this->addParsedBody($request));
    } catch (\Throwable $e) {
        // Here, CORS middleware is not processed
        $response = $this->responder->error(ErrorCode::ERROR_NOT_FOUND, $e->getMessage());
        if ($this->debug) {
            $response = ResponseUtils::addExceptionHeaders($response, $e);
        }
    }
    return $response;
}
```

Here is what I wanted to do to fix the bug:

 - Add a method `getCorsMiddleware` in `SimpleRouter` class
 - Before condition on $this->debug, call this method to get middleware
 - Call middleware::process method to add headers to response

The problem is that CorsMiddleware::process method takes two arguments: `$request` and `$next`. I don't want to pass`$next` here (should be the Api instance).

I see two solutions: 

 - Allow `$next` to be null (but it requires to change the `Middleware` interface, which will possibly break several users APIs.
 - Split `CorsMiddleware::process` code in several functions to allow adding headers to a response anywhere.

Thanks again for your amazing API!

**Comment by mevdschee** [comment]
Okay, so I was reading a bit on this topic and I think we should try to support exceptions thrown in middleware and other handlers while also conforming to PSR 15, see:

https://www.php-fig.org/psr/psr-15/

**Comment by sebj54** [comment]
Oh I see, but it seems more like a huge rework of all middlewares more than just providing a fix for this behavior with CORS headers?

**Comment by mevdschee** [comment]
>Oh I see, but it seems more like a huge rework of all middlewares

No, not really actually. As the code already conforms to [PSR 15](https://www.php-fig.org/psr/psr-15/). It does help you with some of your decisions.

**Comment by surruk51** [comment]
I had a similar problem. I set $config->debug = true because I was getting a 9999 error when trying to use POST to insert an item into a table (tried with two separate tables), but I was getting no X-Debug-Info header in my response. I found at: api.php:11200 a function which handles exceptions, and which does not add X-Debug-Info, but does add X-Exception-Name, X-Exception-Message, X-Exception-File' headers. However, I was still not getting these. I figured it was the CORS filter implemented in javascript fetch() function, so I added these two lines to the function at  api.php:11200 to test that out:            
     $response = $response->withHeader('Access-Control-Expose-Headers','X-Debug-Info, X-Exception-Name, X-Exception-Message, X-Exception-File');
     $response = $response->withHeader('Access-Control-Allow-Headers': 'X-Debug-Info, X-Exception-Name, X-Exception-Message, X-Exception-File');
However, I am still not getting the debug info. SO, I think I am lost. How can I get my debug info and find out what is causing this problem.
